network/p2p/scoring/registry_test.go

```diff
@@ -422,6 +423,63 @@ func TestPersistingInvalidSubscriptionPenalty(t *testing.T) {
 	assert.Equal(t, 0.0, record.Penalty) // penalty should be zero.
 }
 
+// TestPeerSpamPenaltyClusterPrefixed evaluates the application-specific penalty calculation for a node when a spam record is present
+// for cluster-prefixed topics. In the case of an invalid control message notification marked as cluster-prefixed,
+// the application-specific penalty should be reduced by the default reduction factor. This test verifies the accurate computation
+// of the application-specific score under these conditions.
+func TestPeerSpamPenaltyClusterPrefixed(t *testing.T) {
+	peerID := unittest.PeerIdFixture(t)
+	reg, spamRecords := newGossipSubAppSpecificScoreRegistry(
+		t,
+		withStakedIdentity(peerID),
+		withValidSubscriptions(peerID))
+
+	// initially, the spamRecords should not have the peer id.
+	assert.False(t, spamRecords.Has(peerID))
+
+	// since the peer id does not have a spam record, the app specific score should be the max app specific reward, which
+	// is the default reward for a staked peer that has valid subscriptions.
+	score := reg.AppSpecificScoreFunc()(peerID)
+	assert.Equal(t, scoring.MaxAppSpecificReward, score)
+
+	// Report consecutive misbehavior's for the specified peer ID. Two misbehavior's are reported concurrently:
+	// 1. With IsClusterPrefixed set to false, ensuring the penalty applied to the application-specific score is not reduced.
+	// 2. With IsClusterPrefixed set to true, reducing the penalty added to the overall app-specific score by the default reduction factor.
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		reg.OnInvalidControlMessageNotification(&p2p.InvCtrlMsgNotif{
+			PeerID:            peerID,
+			MsgType:           p2pmsg.CtrlMsgGraft,
+			IsClusterPrefixed: false,
+		})
+	}()
+	go func() {
+		defer wg.Done()
+		reg.OnInvalidControlMessageNotification(&p2p.InvCtrlMsgNotif{
+			PeerID:            peerID,
+			MsgType:           p2pmsg.CtrlMsgGraft,
+			IsClusterPrefixed: true,
+		})
+	}()
+	unittest.RequireReturnsBefore(t, wg.Wait, 100*time.Millisecond, "timed out waiting for goroutines to finish")
+
+	// expected penalty should be penaltyValueFixtures().Graft * (1  + clusterReductionFactor)
+	expectedPenalty := penaltyValueFixtures().Graft * (1 + penaltyValueFixtures().ClusterPrefixedPenaltyReductionFactor)
+
+	// the penalty should now be updated in the spamRecords
+	record, err, ok := spamRecords.Get(peerID) // get the record from the spamRecords.
+	assert.True(t, ok)
+	assert.NoError(t, err)
+	assert.Less(t, math.Abs(expectedPenalty-record.Penalty), 10e-3)
+	assert.Equal(t, scoring.InitAppScoreRecordState().Decay, record.Decay)
+	// this peer has a spam record, with no subscription penalty. Hence, the app specific score should only be the spam penalty,
+	// and the peer should be deprived of the default reward for its valid staked role.
+	score = reg.AppSpecificScoreFunc()(peerID)
+	assert.Less(t, math.Abs(expectedPenalty-score), 10e-3)
```

Comment:
How did you choose the tolerance for `assert.Less`? I'm seeing this test pass for 2 orders of magnitude smaller:

```
	assert.Less(t, math.Abs(expectedPenalty-record.Penalty), 10e-5)
	assert.Equal(t, scoring.InitAppScoreRecordState().Decay, record.Decay)
	// this peer has a spam record, with no subscription penalty. Hence, the app specific score should only be the spam penalty,
	// and the peer should be deprived of the default reward for its valid staked role.
	score = reg.AppSpecificScoreFunc()(peerID)
	assert.Less(t, math.Abs(expectedPenalty-score), 10e-5)
```

Reply:
We are compensating for the potential difference in resources between local machine and remote CI instances. Also the difference is very small and does not decrease or increase the effectiveness of the test.